File: dart/lib/acme/application/command/handler/CreateUserHandler.dart
import 'package:dart/acme/application/command/CreateUserCommand.dart';
import 'package:dart/acme/domain/model/user.dart';
import 'package:dart/acme/domain/users.dart';
import 'package:dart/sharedkernel/application/command/command.dart';
import 'package:dart/sharedkernel/application/command/handler/handler.dart';

class CreateUserHandler implements Handler {
  final Users _users;

  CreateUserHandler(this._users);

  @override
  void handle(Command command) {
    if (!(command is CreateUserCommand)) {
      throw Exception('Invalid command type');
    }

    var createUserCommand = (command as CreateUserCommand);

    if (_users.exists(createUserCommand.email)) {
      throw Exception('Current user exists');
    }

    var user = User(
        createUserCommand.email,
        createUserCommand.password,
        createUserCommand.firstName,
        createUserCommand.lastName
    );

    _users.add(user);
  }
}

File: php/src/Acme/Application/Command/Handler/CreateUserHandler.php
<?php

declare(strict_types=1);

namespace Acme\Application\Command\Handler;

use Acme\Application\Command\CreateUserCommand;
use Acme\Domain\Model\User;
use Acme\Domain\Users;
use SharedKernel\Application\Command\Command;
use SharedKernel\Application\Command\Handler\Handler;

class CreateUserHandler implements Handler
{
    /**
     * @var Users
     */
    private $users;

    public function __construct(Users $users)
    {
        $this->users = $users;
    }

    /**
     * @throws \Exception
     */
    public function handle(Command $command): void
    {
        if (!($command instanceof CreateUserCommand)) {
            throw new \RuntimeException('Invalid command type');
        }

        if ($this->users->exists($command->email())) {
            throw new \InvalidArgumentException('Current user exists');
        }

        $user = new User($command->email(), $command->password(), $command->firstName(), $command->lastName());

        $this->users->add($user);
    }
}
